frame/support/procedural/src/benchmark.rs

```diff
@@ -0,0 +1,866 @@
+
+					// consume #[block] tokens
+					block.attrs.remove(k);
+
+					Some(Ok((i, BenchmarkCallDef::Block { block, attr_span: attr.span() })))
+				})
+			} else {
+				None
+			}
+		}).collect::<Result<Vec<_>>>()?;
+		let (i, call_def) = match &call_defs[..] {
+			[(i, call_def)] => (*i, call_def.clone()), // = 1
+			[] => return Err(Error::new( // = 0
+				item_fn.block.brace_token.span,
+				"No valid #[extrinsic_call] or #[block] annotation could be found in benchmark function body."
+			)),
+			_ => return Err(Error::new( // > 1
+				call_defs[1].1.attr_span(),
+				"Only one #[extrinsic_call] or #[block] attribute is allowed per benchmark."
+			)),
+		};
+
+		Ok(BenchmarkDef {
+			params,
+			setup_stmts: Vec::from(&item_fn.block.stmts[0..i]),
+			call_def,
+			verify_stmts: Vec::from(&item_fn.block.stmts[(i + 1)..item_fn.block.stmts.len()]),
+			extra,
+			skip_meta,
+		})
+	}
+}
+
+/// Parses and expands a `#[benchmarks]` or `#[instance_benchmarks]` invocation
+pub fn benchmarks(
+	attrs: TokenStream,
+	tokens: TokenStream,
+	instance: bool,
+) -> syn::Result<TokenStream> {
+	// gather module info
+	let module: ItemMod = syn::parse(tokens)?;
+	let mod_span = module.span();
+	let where_clause = match syn::parse::<Nothing>(attrs.clone()) {
+		Ok(_) => quote!(),
+		Err(_) => syn::parse::<WhereClause>(attrs)?.predicates.to_token_stream(),
+	};
+	let mod_vis = module.vis;
+	let mod_name = module.ident;
+
+	// consume #[benchmarks] attribute by exclusing it from mod_attrs
+	let mod_attrs: Vec<&Attribute> = module
+		.attrs
+		.iter()
+		.filter(|attr| syn::parse2::<keywords::benchmarks>(attr.to_token_stream()).is_err())
+		.collect();
+
+	let mut benchmark_names: Vec<Ident> = Vec::new();
+	let mut extra_benchmark_names: Vec<Ident> = Vec::new();
+	let mut skip_meta_benchmark_names: Vec<Ident> = Vec::new();
+
+	let (_brace, mut content) =
+		module.content.ok_or(syn::Error::new(mod_span, "Module cannot be empty!"))?;
+
+	// find all function defs marked with #[benchmark]
+	let benchmark_fn_metas = content.iter_mut().filter_map(|stmt| {
+		// parse as a function def first
+		let Item::Fn(func) = stmt else { return None };
+
+		// find #[benchmark] attribute on function def
+		let benchmark_attr = func.attrs.iter().find_map(|attr| {
+			let seg = attr.path.segments.last()?;
+			syn::parse::<keywords::benchmark>(seg.ident.to_token_stream().into()).ok()?;
+			Some(attr)
+		})?;
+
+		Some((benchmark_attr.clone(), func.clone(), stmt))
+	});
+
+	// parse individual benchmark defs and args
+	for (benchmark_attr, func, stmt) in benchmark_fn_metas {
+		// parse any args provided to #[benchmark]
+		let attr_tokens = benchmark_attr.tokens.to_token_stream().into();
+		let benchmark_args: BenchmarkAttrs = syn::parse(attr_tokens)?;
+
+		// parse benchmark def
+		let benchmark_def =
+			BenchmarkDef::from(&func, benchmark_args.extra, benchmark_args.skip_meta)?;
+
+		// expand benchmark
+		let expanded = expand_benchmark(
+			benchmark_def.clone(),
```

Comment:
If you put `expand_benchmark` after recording the name sets, then you won't need to clone here.